src/entities/elastic-storage/elastic-storage-processor.js

```diff
@@ -27,66 +76,165 @@ class ElasticStorageProcessor implements StorageProcessor {
 
   networkStartTime: number
 
+  elasticConfig: ElasticConfigType
+
   constructor(
     logger: Logger,
-    elasticNode: string,
+    elasticConfig: ElasticConfigType,
     networkConfig: NetworkConfig,
   ) {
     this.logger = logger
-    this.client = new Client({ node: elasticNode })
+    this.elasticConfig = elasticConfig
+    this.client = new Client({ node: elasticConfig.node })
     this.networkStartTime = networkConfig.startTime()
   }
 
+  indexFor(name: string) {
+    // TOFO: memoize
+    return `${this.elasticConfig.indexPrefix}.${name}`
+  }
+
+  async rollbackTo(height: number) {
+    await sleep(10000)
+    const latestStableChunk = await this.getLatestStableChunk()
+    return this.deleteChunksAfter(Math.min(latestStableChunk, height))
+  }
+
+
+  async esSearch(params: {}) {
+    const resp = await this.client.search(params)
+    const { hits } = resp.body
+    return hits
+  }
+
+  async getLatestStableChunk() {
+    const index = this.indexFor(INDEX_CHUNK)
+    const indexExists = (await this.client.indices.exists({
+      index,
+    })).body
+    if (!indexExists) {
+      return 0
+    }
+    const hits = await this.esSearch({
+      index,
+      allowNoIndices: true,
+      ignoreUnavailable: true,
+      body: {
+        sort: [{ chunk: { order: 'desc' } }],
+        size: 1,
+      },
+    })
+    this.logger.debug('getLatestStableChunk', hits)
+    return hits.total.value > 0 ? hits.hits[0]._source.chunk : 0
+  }
+
+  async deleteChunksAfter(chunk: number) {
+    const resp = await this.client.deleteByQuery({
+      index: this.indexFor(INDEX_POINTER_ALL),
+      body: {
+        query: { range: { _chunk: { gt: chunk } } },
+      },
+    })
+    const deletedDocs = resp.body.total
+    this.logger.info(`deleteChunksAfter(${chunk}), total deleted:${deletedDocs}`, resp)
+  }
+
+  async ensureElasticTemplates() {
+    for (const [name, tmpl] of _.toPairs(ELASTIC_TEMPLATES)) {
+      // eslint-disable-next-line no-await-in-loop
+      const tmplExists = await this.client.indices.existsTemplate({
+        name,
+      })
+      if (!tmplExists.body) {
+        // eslint-disable-next-line no-await-in-loop
+        const resp = await this.client.indices.putTemplate({
+          name,
+          body: tmpl,
+          include_type_name: false,
+        })
+        this.logger.debug(`Put template ${name}`, resp)
+      }
+    }
+  }
+
+  async storeChunk(chunkBody: ChunkBodyType) {
+    return this.client.index({
+      index: this.indexFor(INDEX_CHUNK),
+      id: chunkBody.chunk,
+      body: chunkBody,
+    })
+  }
+
+  async removeUnsealed() {
+    const lastChunk = await this.getLatestStableChunk()
+    this.logger.debug('Remove unsealed blocks after', lastChunk)
+    if (lastChunk > 0) {
+      await this.deleteChunksAfter(lastChunk)
+    }
+  }
+
   async genesisLoaded() {
+    await this.ensureElasticTemplates()
+    await this.removeUnsealed()
```

Comment:
Let's add a new function to the general storage-processor interface, something like `onLaunch()`, and move this functionality there.